### PR TITLE
drivers: add radio driver for nrf52840

### PR DESCRIPTION
**Pull request by dnahm**
This PR includes the radio driver for nrf52 controller.
Bug: sometimes it doesn't answer a ping6.

### CODE REVIEW

**Comment by OlegHahm** [comment]
Nice job! 👍 

Have you tried to figure out at which layer the response to the ping gets lost? Is the ping properly received and passed up the stack?

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+/**
+ * @brief   Set radio into idle (DISABLED) state
+ */
+static void go_idle(void)
+{
+    /* set device into basic disabled state */
+    NRF_RADIO->EVENTS_DISABLED = 0;
+    NRF_RADIO->TASKS_DISABLE = 1;
+    while (NRF_RADIO->EVENTS_DISABLED == 0) {}
```
Comment (OlegHahm) [inline cpu/nrf52/radio/nrf802154/nrf802154.c]: Maybe for debugging purposes add a counter and an assertion in order to prevent infinite loops.

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+    }
+    else if (idle_state == STATE_OFF) {
+        NRF_RADIO->POWER = 0;
+    }
+}
+
+static void set_chan(uint16_t chan)
+{
+    assert((chan >= IEEE802154_CHANNEL_MIN) && (chan <= IEEE802154_CHANNEL_MAX));
+    NRF_RADIO->FREQUENCY = (chan - 10) * 5;
```
Comment (OlegHahm) [inline cpu/nrf52/radio/nrf802154/nrf802154.c]: Here and other locations: please use macro defines for constants instead of literals wherever possible.

File: examples/default/main.c [hunk 3]
```diff
@@ -38,6 +38,36 @@
...
+    printf("STATE: 0x%08x\n", (int)NRF_RADIO->STATE);
+
+    printf("FREQ:  0x%08x\n", (int)NRF_RADIO->FREQUENCY);
+    printf("PCNF0: 0x%08x\n", (int)NRF_RADIO->PCNF0);
+    printf("PCNF1: 0x%08x\n", (int)NRF_RADIO->PCNF1);
+    printf("SFD:   0x%08x\n", (int)NRF_RADIO->SFD);
+    printf("TXPOW: 0x%08x\n", (int)NRF_RADIO->TXPOWER);
+
+    printf("MHRMATCHCONF: 0x%08x\n", (int)NRF_RADIO->MHRMATCHCONF);
+    printf("MHRMATCHMAS:  0x%08x\n", (int)NRF_RADIO->MHRMATCHMAS);
```
Comment (OlegHahm) [inline examples/default/main.c]: These NRF specific lines shouldn't go into the default example.

**Comment by miri64** [comment]
Ping @dnahm?

**Comment by dnahm** [comment]
The bug is fixed. ACK response is included.
But currently I don't know, if the ACK response works correctly.

File: core/include/debug.h [hunk 4]
```diff
@@ -52,12 +52,7 @@ extern "C" {
 #include "cpu_conf.h"
 #define DEBUG_PRINT(...) \
     do { \
-        if ((sched_active_thread == NULL) || (sched_active_thread->stack_size > THREAD_EXTRA_STACKSIZE_PRINTF)) { \
```
Comment (smlng) [inline core/include/debug.h]: looks unrelated?

Comment (dnahm) [reply core/include/debug.h]: @haukepetersen Can you verify this?

Comment (haukepetersen) [reply core/include/debug.h]: This is indeed unrelated. I actually don't remember the reason for this if here (except that is mostly very annoying), but I think there were platforms that could not handle printf in interrupt context -> which poses the question again, why the custom printf is not executed, but instead a puts with a fairly long string is... But this discussion is independent of this PR, so just take this change out.

File: cpu/nrf52/include/nrf802154.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#ifndef NRF802154_H
+#define NRF802154_H
+
+#include "net/netdev.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
```
Comment (haukepetersen) [inline cpu/nrf52/include/nrf802154.h]: one newline too much here

File: cpu/nrf52/include/nrf802154.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * @brief   Export the netdev device descriptor
+ */
+extern netdev_ieee802154_t nrf802154_dev;
+
+/**
+ * @brief   Reference to the netdev driver interface
+ */
```
Comment (haukepetersen) [inline cpu/nrf52/include/nrf802154.h]: this comment can be removed

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+ * Copyright (C) 2017 Freie Universität Berlin
+ *               2017 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_nrf52
```
Comment (haukepetersen) [inline cpu/nrf52/radio/nrf802154/nrf802154.c]: should be `drivers_nrf52_802154`, right?

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+ {
+     /* reset RX buffer and listen for new packets */
+     rxbuf[0] = 0;
+     NRF_RADIO->TASKS_START = 1;
+ }
+
+/**
+ * @brief   Switch on the radio
+ */
+static void power_on(void)
```
Comment (haukepetersen) [inline cpu/nrf52/radio/nrf802154/nrf802154.c]: you could make this function explicitly inline (but the compile does this probably anyway, so its rather one for the eye...)

File: cpu/nrf52/radio/nrf802154/nrf802154.c (hunk 2 above)
Comment (haukepetersen) [reply cpu/nrf52/radio/nrf802154/nrf802154.c]: In this place I think magic numbers are ok (as its really hard to give them good names here. But you could add a short comment on how channels and frequency values are mapped

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+    /* and set some fitting configuration */
+    NRF_RADIO->PCNF0 = ((8 << RADIO_PCNF0_LFLEN_Pos) |
+                        (RADIO_PCNF0_PLEN_32bitZero << RADIO_PCNF0_PLEN_Pos) |
+                        (RADIO_PCNF0_CRCINC_Include << RADIO_PCNF0_CRCINC_Pos));
+    NRF_RADIO->PCNF1 = IEEE802154_FRAME_LEN_MAX;
+    /* set start frame delimiter */
+    NRF_RADIO->SFD = IEEE802154_SFD;
+    /* set MHR filters */
+    NRF_RADIO->MHRMATCHCONF = 0;
+    NRF_RADIO->MHRMATCHMAS = 0xff0007ff;
```
Comment (haukepetersen) [inline cpu/nrf52/radio/nrf802154/nrf802154.c]: This config step might well deserve a little more doc

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+
+    /* specify the length of the package. */
+    txbuf[0] = pos + 1;
+
+    enable_tx();
+
+    /* trigger the actual transmission */
+    DEBUG("[nrf802154] send: putting %i byte into the ether\n", pos);
+
+    /* @todo: how about CCA, retransmits, and ACKs? */
```
Comment (haukepetersen) [inline cpu/nrf52/radio/nrf802154/nrf802154.c]: I would propose to add some comprehensive doc and a TODO list into the module description (`nrf802154.h`). Maye have look at the `* ## Implementation Status ##` section in  `gcoap.h` as example. When documented there, you can safely remove the comment above

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+        case NETOPT_TX_POWER:
+            assert(value_len == sizeof(int16_t));
+            set_txpower(*((int16_t *)value));
+            return sizeof(int16_t);
+
+        default:
+            return netdev_ieee802154_set((netdev_ieee802154_t *)dev,
+                                         opt, value, value_len);
+    }
+
```
Comment (haukepetersen) [inline cpu/nrf52/radio/nrf802154/nrf802154.c]: newline...

File: cpu/nrf52/radio/nrf802154/nrf802154.c [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+        if ((nrf802154_dev.netdev.event_callback) &&
+            (NRF_RADIO->CRCSTATUS == 1) &&
+            netdev_ieee802154_dst_filter(&nrf802154_dev, &rxbuf[1])) {
+
+            DEBUG("isr-rx: packet valid -> processing\n");
+
+            txbuf[0] = 5;
+            txbuf[1] = 2;
+            txbuf[2] = 0;
+            txbuf[3] = rxbuf[3];
```
Comment (haukepetersen) [inline cpu/nrf52/radio/nrf802154/nrf802154.c]: this seems dangerous to me. Are we sure, that there is definitely no TX operation in process at this point which could be corrupted by writing in `txbuf`?

Comment (haukepetersen) [reply cpu/nrf52/radio/nrf802154/nrf802154.c]: As fix, we could add a dedicated TX buffer for ACK messages.

File: drivers/netdev_ieee802154/netdev_ieee802154.c [hunk 13]
```diff
@@ -36,8 +36,8 @@ static int _get_iid(netdev_ieee802154_t *dev, eui64_t *value, size_t max_len)
 
     if (dev->flags & NETDEV_IEEE802154_SRC_MODE_LONG) {
         addr_len = IEEE802154_LONG_ADDRESS_LEN;
-        addr = dev->long_addr;
-    }
+        addr = dev->long_addr
+;    }
```
Comment (haukepetersen) [inline drivers/netdev_ieee802154/netdev_ieee802154.c]: whats up with this semicolon?

File: cpu/nrf52/radio/nrf802154/nrf802154.c (hunk 12 above)
Comment (haukepetersen) [reply cpu/nrf52/radio/nrf802154/nrf802154.c]: One more thing: the sending of ACK messages should be conditional, depending if the AUTOACK flag was set. But I would be ok if you add this as a TODO to the implementation status doc.

**Comment by haukepetersen** [comment]
Looking good, will test at tonight's Hack'n'Ack. Could you please rebase until then? Thx

**Comment by smlng** [comment]
need rebase

**Comment by dnahm** [comment]
I can't find the bug.

**Comment by haukepetersen** [comment]
@dnahm what is the status on this PR? Anything blocking from your perspective?

**Comment by dnahm** [comment]
Not all features are implemented (like Acknowledgement) and sometimes it doesn't work correctly.

**Comment by haukepetersen** [comment]
I see, but do you plan on working on this anytime soon?

**Comment by miri64** [comment]
Ping (@PeterKietzmann maybe?)

**Comment by haukepetersen** [comment]
I will take over this PR, can't promise on a date though...

**Comment by miri64** [comment]
> I will take over this PR, can't promise on a date though...

Then let it close as memo for now.
